drgn/helpers/linux/wait.py

```diff
@@ -0,0 +1,77 @@
+# Copyright (c) 2023, Oracle and/or its affiliates.
+# SPDX-License-Identifier: LGPL-2.1-or-later
+
+"""
+Waitqueue
+--------------
+
+The ``drgn.helpers.linux.waitq`` module provides helpers for working with the
+wait-queue mechanism, implemented in :linux:`include/linux/wait.h`
+(``struct wait_queue_head`` and ``struct wait_queue_entry``).
+"""
+
+from typing import Iterator
+
+from drgn import Object, cast
+from drgn.helpers.linux.list import list_empty, list_for_each_entry
+
+__all__ = (
+    "waitqueue_active",
+    "waitqueue_for_each_entry",
+    "waitqueue_for_each_task",
+)
+
+
+def _get_wait_queue_head(waitq: Object) -> Object:
+    # Linux kernel commit 2055da97389a ("sched/wait:
+    # Disambiguate wq_entry->task_list and wq_head->task_list
+    # naming") (in v4.13) renamed the task_list member to head.
+    try:
+        return waitq.head
+    except AttributeError:
+        return waitq.task_list
+
+
+def waitqueue_active(waitq: Object) -> bool:
+    """
+    Check if a waitqueue is empty or not.
+
+    :param waitq: ``wait_queue_head_t *``
+    :return: True if waitqueue has waiters, False otherwise
+    """
+    head = _get_wait_queue_head(waitq)
+    return not list_empty(head.address_of_())
+
+
+def waitqueue_for_each_entry(waitq: Object) -> Iterator[Object]:
+    """
+    Iterate over all entries in a given wait-queue.
+
+    :param waitq: ``wait_queue_head_t *``
+    :return: Iterator of ``struct wait_queue_entry *`` or ``struct __wait_queue *`` objects.
```

Comment:
Since you're using the typedef names now, we should document it that way, and also specify why it can be different.

```suggestion
    :return: Iterator of ``wait_queue_entry_t *`` or ``wait_queue_t *`` objects
        depending on the kernel version.
```

Reply:
Agree. Although the comment later in the function gives a summary of the change, I have further explained the relation between these 2 typedefs. Please let me know if it looks good or needs any change.